rust/worker/src/assignment/rendezvous_hash.rs

```diff
@@ -0,0 +1,173 @@
+// This implementation mirrors the rendezvous hash implementation
+// in the go and python services.
+// The go implementation is located go/internal/utils/rendezvous_hash.go
+// The python implementation is located chromadb/utils/rendezvous_hash.py
+
+use crate::errors::{ChromaError, ErrorCodes};
+use std::io::Cursor;
+use thiserror::Error;
+
+use murmur3::murmur3_x64_128;
+
+/// A trait for hashing a member and a key to a score.
+trait Hasher {
+    fn hash(&self, member: &str, key: &str) -> Result<u64, AssignmentError>;
+}
+
+/// Error codes for assignment
+#[derive(Error, Debug)]
+enum AssignmentError {
+    #[error("Cannot assign empty key")]
+    EmptyKey,
+    #[error("No members to assign to")]
+    NoMembers,
+    #[error("Error hashing member")]
+    HashError,
+}
+
+impl ChromaError for AssignmentError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            AssignmentError::EmptyKey => ErrorCodes::InvalidArgument,
+            AssignmentError::NoMembers => ErrorCodes::InvalidArgument,
+            AssignmentError::HashError => ErrorCodes::Internal,
+        }
+    }
+}
+
+/// Assign a key to a member using the rendezvous hash algorithm.
+/// # Arguments
+/// - key: The key to assign.
+/// - members: The members to assign to.
+/// - hasher: The hasher to use.
+/// # Returns
+/// The member that the key was assigned to.
+/// # Errors
+/// - If the key is empty.
+/// - If there are no members to assign to.
+/// - If there is an error hashing a member.
+/// # Notes
+/// This implementation mirrors the rendezvous hash implementation
+/// in the go and python services.
+fn assign<H: Hasher>(
+    key: &str,
+    members: impl IntoIterator<Item = impl AsRef<str>>,
+    hasher: &H,
+) -> Result<String, AssignmentError> {
+    if key.is_empty() {
+        return Err(AssignmentError::EmptyKey);
+    }
+
+    let mut iterated = false;
+    let mut max_score = u64::MIN;
+    let mut max_member = None;
+
+    for member in members {
+        if !iterated {
```

Comment:
Note to myself: we need this because `count()` on a rust `Iterator` consumes the iterator and `size_hint()` explicitly states that iterators don't have to implement it correctly. So there's no way to check the size of `members` at the top of the function.